### PR TITLE
Allow `Maybe_null` jkinds in more places

### PR DESCRIPTION
**Pull request by dkalinichenko-js**
Make most jkinds created from sort variables `Maybe_null`. This allows types like `or_null` in function arguments, function results, record fields and so on. This excludes:
* Unannotated type parameters for compatibility reasons
* Jkinds created by `transl_simple_type ~new_var_jkind:Sort`
  - Changing any of those to accept `or_null` breaks compatibility with existing code due to more permissive type inference in signatures
* Array elements, to prohibit `float or_null array`

Accept `value_or_null` for the V1 safety check, for the separability check, for tuple elements, for structure elements, and for polymorphic variant fields.

Require array type parameters to be non-null.

Hide `value_or_null` and `any_non_null` from users if the layouts extension is below `Alpha`, replacing them with `value` or `any` in the compiler output.

Convert `any` annotations to `any_non_null` if the layouts extension is below `Alpha` for compatibility with existing code using arrays.

Split some tests into alpha and non-alpha versions due to different printing behavior.

Tests in `typing-layouts-or-null` cover some changes in this PR, but not all. More tests will be added in the PR implementing the `or_null` type.

### CODE REVIEW

File: ocaml/typing/jkind.mli [hunk 1]
```diff
@@ -301,15 +299,21 @@ val add_mode_crossing : t -> t
...
+val of_new_sort_var : why:History.concrete_creation_reason -> t * sort
 
 (** Create a fresh sort variable, packed into a jkind. *)
-val of_new_sort : why:History.concrete_default_creation_reason -> t
+val of_new_sort : why:History.concrete_creation_reason -> t
+
+(** Same as [of_new_sort_var], but the jkind is lowered to [Non_null].
+    Defaulting the sort variable produces exactly [value].  *)
+val of_new_default_sort_var :
+  why:History.concrete_default_creation_reason -> t * sort
```
[the remark this answers is not in the document]
Comment (dkalinichenko-js) [reply ocaml/typing/jkind.mli:312]: This isn't exclusive to `non_null`: after implementing separability, this jkind would also need to be `separable`, the default non-top choice for separability. So I don't want to put `non_null` in the name.

Perhaps the right naming is `concrete_legacy_creation_reason`, as those sort variables mimic "legacy" OCaml values? Suggestions are welcome.

[the remark this answers is not in the document]
Comment (goldfirere) [reply ocaml/typing/jkind.mli:312]: Yes, I’m happy with legacy. It makes me think about predetermined choices, not dynamically running out of options and then choosing a default. Thanks.

File: ocaml/testsuite/tests/typing-layouts-arrays/basics_alpha.ml
[the remark this answers is not in the document]
Comment (dkalinichenko-js) [reply ocaml/testsuite/tests/typing-layouts-arrays/basics_alpha.ml:1]: Done.

File: ocaml/typing/ctype.ml [hunk 2]
```diff
@@ -2277,6 +2277,12 @@ let type_sort ~why env ty =
   | Ok _ -> Ok sort
   | Error _ as e -> e
 
+let type_non_null_sort ~why env ty =
+  let jkind, sort = Jkind.of_new_default_sort_var ~why in
```
[the remark this answers is not in the document]
Comment (dkalinichenko-js) [reply ocaml/typing/ctype.ml]: Changed everything to `legacy`.

**Pull request merged**